### PR TITLE
[App Config] `az appconfig replica`: New command group to support geo-replication

### PR DESCRIPTION
**Pull request by shenmuxiaosen**
Description
App Configuration now supports geo-replication feature. This PR now supports adding new replicas, listing them and deleting the replicas.

Changes in this feature -

- Add new command group ```replica``` - ```az appconfig replica```.
- Add new command ```az appconfig replica create``` which will create replica.
- Add new command ```az appconfig replica list``` which will list replicas of App Configuration.
- Add new command ```az appconfig replica show``` which will show the properties of a replica of an App Configuration.
- Add new command ```az appconfig replica delete``` which will delete a replica of an App Configuration.

Testing Guide

- Create a new App Configuration.
```az appconfig create -g MyResourceGroup -n MyAppConfiguration -l westus --sku Standard```

- Create a new replica for the App Configuration.
```az appconfig replica create -g MyResourceGroup -n MyReplicaName -s MyAppConfiguration -l westus```

- List all replicas of App Configurations.
```az appconfig replica list -s MyAppConfiguration```

- Show properties of a replica of App Configuration.
```az appconfig replica show -n MyReplicaName -s MyAppConfiguration```

- Delete a replica of App Configuration.
```az appconfig replica delete -n MyReplicaName -s MyAppConfiguration```

**History Notes**
[App Config] `az appconfig replica`: New command group to support geo-replication

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
App Config

File: src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py [hunk 1]
```diff
@@ -10,16 +10,29 @@
 logger = get_logger(__name__)
 
 
+STABLE_API_VERSION = '2022-05-01'
+REPLICATION_API_VERSION = '2022-03-01-preview'
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py:14]: The guideline for multiapi SDKs is to define a new resource type for appconfiguration in https://github.com/Azure/azure-cli/blob/21d07dc528d283030e5911243c24bef2d2c18d88/src/azure-cli-core/azure/cli/core/profiles/_shared.py#L43

and then specify the API version for different operations in each profile https://github.com/Azure/azure-cli/blob/21d07dc528d283030e5911243c24bef2d2c18d88/src/azure-cli-core/azure/cli/core/profiles/_shared.py#L152

for example, `azure-mgmt-keyvault` uses `2021-04-01-preview` as default API version but uses `2022-07-01` for `vaults_operations`: https://github.com/Azure/azure-cli/blob/21d07dc528d283030e5911243c24bef2d2c18d88/src/azure-cli-core/azure/cli/core/profiles/_shared.py#L180-L182

So appconfiguration's definition could be something like
```
        ResourceType.MGMT_APPCONFIGURATION: SDKProfile('2022-05-01', {
            'replicas': '2022-03-01-preview'
        }),
```

File: src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py [hunk 2]
```diff
@@ -10,16 +10,29 @@
...
+
+
 def get_appconfig_service_client(cli_ctx, api_version=None):
     ''' Returns the client for managing configuration stores.'''
     from azure.mgmt.appconfiguration import AppConfigurationManagementClient
-    return get_mgmt_service_client(cli_ctx, AppConfigurationManagementClient, api_version=api_version)
+    api_version = STABLE_API_VERSION if not api_version else api_version
+    client = get_mgmt_service_client(cli_ctx, AppConfigurationManagementClient, api_version=api_version)
+    client.api_version = api_version
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py:22]: By defining API version for each operation in cli core, you don't need to specify `api_version` here. `get_mgmt_service_client` will smartly find the exact version.

File: src/azure-cli/azure/cli/command_modules/appconfig/custom.py [hunk 3]
```diff
@@ -243,6 +246,43 @@ def regenerate_credential(cmd, client, name, id_, resource_group_name=None):
...
+                               config_store_name=store_name,
+                               replica_name=name,
+                               replica_creation_parameters=replica_creation_params)
+
+
+def delete_replica(cmd, client, store_name, name, resource_group_name=None, yes=False):
+    if resource_group_name is None:
+        resource_group_name, _ = resolve_store_metadata(cmd, store_name)
+
+    user_confirmation("Are you sure you want to delete the replica '{}' for the App Configuration '{}'".format(name, store_name), yes)
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/appconfig/custom.py:280]: You can add `confirmation=True` in command definition instead

File: src/azure-cli/azure/cli/command_modules/appconfig/commands.py [hunk 4]
```diff
@@ -64,6 +72,12 @@ def get_custom_sdk(custom_module, client_factory, table_transformer):
         g.show_command('show', 'show_configstore')
         g.show_command('show-deleted', 'show_deleted_configstore', table_transformer=deleted_configstore_output_format)
 
+    with self.command_group('appconfig replica', configstore_replica_util, is_preview=True) as g:
+        g.command('list', 'list_replica')
+        g.command('create', 'create_replica')
+        g.command('delete', 'delete_replica')
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/appconfig/commands.py:78]:
```suggestion
        g.command('delete', 'delete_replica', confirmation=True)
```

**Review by avanigupta — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py [hunk 5]
```diff
@@ -10,16 +10,29 @@
...
+    return client
 
 
 def cf_configstore(cli_ctx, *_):
     return get_appconfig_service_client(cli_ctx).configuration_stores
 
 
+def cf_replicas(cli_ctx, *_):
+    client = get_appconfig_service_client(cli_ctx, REPLICATION_API_VERSION).replicas
+    client.api_version = REPLICATION_API_VERSION
```
[the remark this answers is not in the document]
Comment (shenmuxiaosen) [reply src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py]: Removed.

File: src/azure-cli/azure/cli/command_modules/appconfig/commands.py (hunk 4 above)
Comment (shenmuxiaosen) [reply src/azure-cli/azure/cli/command_modules/appconfig/commands.py]: Fixed.

File: src/azure-cli/azure/cli/command_modules/appconfig/custom.py (hunk 3 above)
Comment (shenmuxiaosen) [reply src/azure-cli/azure/cli/command_modules/appconfig/custom.py]: Fixed.

File: src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py (hunk 2 above)
Comment (shenmuxiaosen) [reply src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py]: Removed the unnecessary input argument.

File: src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py (hunk 1 above)
Comment (shenmuxiaosen) [reply src/azure-cli/azure/cli/command_modules/appconfig/_client_factory.py]: Thanks for the detailed explanation. Fixed as suggested.

**Comment by shenmuxiaosen** [comment]
@evelyn-ys Can you help to take another look? Thanks.

**Review by evelyn-ys — APPROVED** [review]

**Review by zhoxing-ms — APPROVED** [review]

**Comment by zhoxing-ms** [comment]
> Add new command group replica - az appconfig replica.
Add new command az appconfig replica create which will create replica.
Add new command az appconfig replica list which will list replicas of App Configuration.
Add new command az appconfig replica show which will show the properties of a replica of an App Configuration.
Add new command az appconfig replica delete which will delete a replica of an App Configuration.

Please write the specific description of the new commands under `[History Notes]`, so that they can be written into the history notes.

**Comment by zhoxing-ms** [comment]
Since `az appconfig replica` is only a simple command group and only contains CRUD operations, so I think it is OK to write only the description of the command group

**Pull request merged**